src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs

```diff
@@ -606,5 +606,355 @@ class Program
+        Source2 s2 = new Source2();
+        Ultimate u = b switch { false => s1, true => s2 };
+        Console.Write(u);
+        u = b switch { false => s1, true => s2 };
+        Console.Write(u);
+    }
+}
+class Ultimate
+{
+    private readonly string Value;
+    public Ultimate(string value) => Value = value;
+    public override string ToString() => Value;
+    public static implicit operator Ultimate(Target t) => new Ultimate(t.ToString());
+}
+class Target
+{
+    private readonly string Value;
+    public Target(string value) => Value = value;
+    public override string ToString() => Value;
+}
+class Source1
+{
+    public static implicit operator Target(Source1 self) => new Target(""Source1 "");
+}
+class Source2
+{
+    public static implicit operator Target(Source2 self) => new Target(""Source2 "");
+}
+";
+            var compilation = CreateCompilation(source, options: TestOptions.DebugExe.WithNullableContextOptions(NullableContextOptions.Disable));
+            compilation.VerifyDiagnostics(
+                // (11,42): error CS0029: Cannot implicitly convert type 'Source1' to 'Ultimate'
+                //         Ultimate u = b switch { false => s1, true => s2 };
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "s1").WithArguments("Source1", "Ultimate").WithLocation(11, 42),
+                // (11,54): error CS0029: Cannot implicitly convert type 'Source2' to 'Ultimate'
+                //         Ultimate u = b switch { false => s1, true => s2 };
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "s2").WithArguments("Source2", "Ultimate").WithLocation(11, 54),
+                // (13,33): error CS0029: Cannot implicitly convert type 'Source1' to 'Ultimate'
+                //         u = b switch { false => s1, true => s2 };
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "s1").WithArguments("Source1", "Ultimate").WithLocation(13, 33),
+                // (13,45): error CS0029: Cannot implicitly convert type 'Source2' to 'Ultimate'
+                //         u = b switch { false => s1, true => s2 };
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "s2").WithArguments("Source2", "Ultimate").WithLocation(13, 45));
+        }
+
+        [Fact]
+        public void SwitchExpressionDiscardedWithNoNaturalType()
+        {
+            var source = @"
+class Program
+{
+    static void Main(string[] args)
+    {
+        _ = true switch { true => 1, false => string.Empty };
+    }
+}
+";
+            var compilation = CreateCompilation(source, options: TestOptions.DebugExe.WithNullableContextOptions(NullableContextOptions.Disable));
+            compilation.VerifyDiagnostics(
+                // (6,18): error CS8506: No best type was found for the switch expression.
+                //         _ = true switch { true => 1, false => string.Empty };
+                Diagnostic(ErrorCode.ERR_SwitchExpressionNoBestType, "switch").WithLocation(6, 18));
+        }
     }
 }
```

Comment:
Suggested to add test

`Target t = c switch { 1 => TargetSubtype, 2 => source };`

where U.D. conversion from source to Target

Reply:
can both conversations get stacked? e.g.: UD from source to the othet arm's type and from that type to Target with no direct conversation from source to Target

Reply:
That is exactly what the "Suggested to add test" is asking to show (that yes, it would work).

---
In reply to: [301891848](https://github.com/dotnet/roslyn/pull/37052#discussion_r301891848) [](ancestors = 301891848)